ci-scripts/yocto.groovy

```diff
@@ -0,0 +1,187 @@
+        stage("Fetch sources") {
+            vagrant("/vagrant/cookbook/yocto/fetch-sources-for-recipes.sh ${yoctoDir} ${bitbake_image}")
+        }
+    }
+
+    stage("Bitbake ${imageName}") {
+        vagrant("/vagrant/cookbook/yocto/build-images.sh ${yoctoDir} ${imageName}")
+        if (dev) {
+            vagrant("/vagrant/cookbook/yocto/build-images.sh ${yoctoDir} ${imageName}-dev")
+        }
+    }
+
+    stage("Build SDK ${imageName}") {
+        vagrant("/vagrant/cookbook/yocto/build-sdk.sh ${yoctoDir} ${imageName}")
+        if (dev) {
+            vagrant("/vagrant/cookbook/yocto/build-sdk.sh ${yoctoDir} ${imageName}-dev")
+        }
+    }
+}
+
+void archiveCache(String yoctoDir, boolean archive, String archivePath) {
+    if (archive && archivePath?.trim()) {
+        stage("Archive cache") {
+            vagrant("rsync -trpg ${yoctoDir}/build/downloads/ ${archivePath}/downloads/")
+            vagrant("rsync -trpg ${yoctoDir}/build/sstate-cache/ ${archivePath}/sstate-cache")
+        }
+    }
+}
+
+void archiveArtifacts(String yoctoDir, String suffix) {
+    stage("Archive artifacts") {
+        String artifactDir = "artifacts_${suffix}"
+
+        sh "rm -rf ${artifactDir}"
+        sh "mkdir ${artifactDir}"
+
+        // Copy images and SDK to the synced directory
+        vagrant("/vagrant/ci-scripts/copy_to_archive ${yoctoDir}/build /vagrant/${artifactDir}")
+
+        // And save them in Jenkins
+        archiveArtifacts "${artifactDir}/**"
+    }
+}
+
+void buildWithLayer(String variant_name, String bitbake_image, String layer, String layerPath) {
+    // Store the directory we are executed in as our workspace.
+    String yoctoDir = "/home/yoctouser/pelux_yocto"
+    String manifest = "pelux.xml"
+
+    // We can't do "checkout scm" here since pelux-manifests is not the
+    // canonical repo here.
+    sh "git submodule update --init"
+    try {
+        // Initialize vagrant and repo manifest
+        startVagrant()
+        repoInit(manifest)
+
+        replaceLayer(yoctoDir, layer, layerPath)
+
+        // Setup yocto
+        String templateConf="${yoctoDir}/sources/meta-pelux/conf/variant/${variant_name}"
+        setupBitbake(yoctoDir, templateConf)
+        setupCache(yoctoDir, env.YOCTO_CACHE_URL)
+
+        // Build the images
+        try {
+            buildImageAndSDK(yoctoDir, bitbake_image)
+        } finally { // Archive cache even if there were errors.
+            boolean archive = env.ARCHIVE_CACHE == "true"
+            archiveCache(yoctoDir, archive, env.YOCTO_CACHE_ARCHIVE_PATH)
+        }
+    } finally {
+        shutdownVagrant()
+    }
+}
+
+void replaceLayer(String yoctoDir, String layerName, String newPath) {
+    vagrant("rm -rf ${yoctoDir}/sources/${layerName}")
+    sh "cp -R ${newPath} ${layerName}"
+    vagrant("mv /vagrant/${layerName} ${yoctoDir}/sources/")
+}
+
+void buildManifest(String variant_name, String bitbake_image) {
```

Comment:
Wouldn't it make sense to only have one function which would do both instead of copy&pasting most of the code?

Reply:
I had that at first, but the code became too cluttered and hard to read. And since most of its steps (except the parts that differ) are just calls to the same functions, I figured this was fine.

Reply:
Hm ok I see. Ok.